fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,413 @@
+    wopt(L!("query"), woption_argument_t::no_argument, 'q'),
+    wopt(L!("copy"), woption_argument_t::no_argument, 'c'),
+    wopt(L!("details"), woption_argument_t::no_argument, 'D'),
+    wopt(L!("no-details"), woption_argument_t::no_argument, NO_METADATA_SHORT),
+    wopt(L!("verbose"), woption_argument_t::no_argument, 'v'),
+    wopt(L!("handlers"), woption_argument_t::no_argument, 'H'),
+    wopt(L!("handlers-type"), woption_argument_t::required_argument, 't'),
+];
+
+/// Parses options to builtin function, populating opts.
+/// Returns an exit status.
+fn parse_cmd_opts(
+    opts: &mut FunctionsCmdOpts,
+    optind: &mut usize,
+    argv: &mut [&wstr],
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+) -> Option<c_int> {
+    let cmd = L!("function");
+    let print_hints = false;
+    let mut w = wgetopter_t::new(SHORT_OPTIONS, LONG_OPTIONS, argv);
+    while let Some(opt) = w.wgetopt_long() {
+        match opt {
+            'v' => opts.verbose = true,
+            'e' => opts.erase = true,
+            'D' => opts.report_metadata = true,
+            NO_METADATA_SHORT => opts.no_metadata = true,
+            'd' => {
+                opts.description = w.woptarg.unwrap().to_owned();
+            }
+            'n' => opts.list = true,
+            'a' => opts.show_hidden = true,
+            'h' => opts.print_help = true,
+            'q' => opts.query = true,
+            'c' => opts.copy = true,
+            'H' => opts.handlers = true,
+            't' => {
+                opts.handlers = true;
+                opts.handlers_type = w.woptarg.unwrap().to_owned();
+            }
+            ':' => {
+                builtin_missing_argument(parser, streams, cmd, argv[w.woptind - 1], print_hints);
+                return STATUS_INVALID_ARGS;
+            }
+            '?' => {
+                builtin_unknown_option(parser, streams, cmd, argv[w.woptind - 1], print_hints);
+                return STATUS_INVALID_ARGS;
+            }
+            other => {
+                panic!("Unexpected retval from wgetopt_long: {}", other);
+            }
+        }
+    }
+
+    *optind = w.woptind;
+    STATUS_CMD_OK
+}
+
+pub fn functions(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    args: &mut [&wstr],
+) -> Option<c_int> {
+    let cmd = args[0];
+
+    let mut opts = FunctionsCmdOpts::default();
+    let mut optind = 0;
+    let retval = parse_cmd_opts(&mut opts, &mut optind, args, parser, streams);
+    if retval != STATUS_CMD_OK {
+        return retval;
+    }
+    // Shadow our args with the positionals
+    let args = &args[optind..];
+
+    if opts.print_help {
+        builtin_print_error_trailer(parser, streams, cmd);
+        return STATUS_CMD_OK;
+    }
+
+    let describe = !opts.description.is_empty();
+    if [describe, opts.erase, opts.list, opts.query, opts.copy]
+        .into_iter()
+        .filter(|b| *b)
+        .count()
+        > 1
```

Comment:
This seemed like the cleanest way to do this sort of "only one" check. The C++ just added the bools together, that's possible here but would require an `as i32` cast for every bool.